New Note Frequency Goodie

sublinear opened the pull request:
**What does your Instant Answer do?**

Calculates the frequency in hertz of any musical note.

**What problem does your Instant Answer solve (Why is it better than organic links)?**

The typical search results are static tables hosted on aging websites for a limited selection of tuning frequencies. This instant answer calculates the frequency for any note in any tuning.

**What is the data source for your Instant Answer? (Provide a link if possible)**

Goodie perl script.

**Why did you choose this data source?**

Because it's a goodie, and not a spice.

**Are there any other alternative (better) data sources?**

No.

**What are some example queries that trigger this Instant Answer?**

"note frequency of c5", "frequency of note c5 432", "notefreq g#6"

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**

Musicians and educators.

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**

No.

**Which existing Instant Answers will this one supercede/overlap with?**

None.

**Are you having any problems? Do you need our help with anything?**

No. I don't think so.

**Where did you hear about DuckDuckHack? (For first time contributors)**

DuckDuckGo homepage.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**

![screenshot from 2015-11-06 15 43 27](https://cloud.githubusercontent.com/assets/9429614/11009942/dedf6c0a-84a0-11e5-8a7e-2e6b77428c14.png)

## Checklist
Please place an 'X' where appropriate.

```
- [X] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
- [X] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
- [X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
- [X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
- [X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 10
        - [X] Google Chrome
        - [X] Firefox
        - [] Opera
        - [] IE 11
        - [] Microsoft edge
        
    - Windows 8
        - [] Google Chrome
        - [] Firefox
        - [] Opera
        - [] IE 10

    - Windows 7
        - [] Google Chrome
        - [] Firefox
        - [] Opera
        - [] IE 8
        - [] IE 9
        - [] IE 10

    - Windows XP
        - [] IE 7
        - [] IE 8

    - Linux
        - [X] Google Chrome
        - [X] Firefox
        
    - Mac OSX
        - [] Google Chrome
        - [] Firefox
        - [] Opera
        - [] Safari

    - iOS (iPhone)
        - [] Safari Mobile
        - [] Google Chrome
        - [] Opera

    - iOS (iPad)
        - [] Safari Mobile
        - [] Google Chrome
        - [] Opera

    - Android
        - [] Firefox
        - [] Native Browser
        - [X] Google Chrome
        - [X] Opera

```

-----
IA Page: https://duck.co/ia/view/note_frequency

daxtheduck commented:
@sublinear Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please send an email to <quackslack@duckduckgo.com>.

>More Info: https://duck.co/duckduckhack/submission_and_review

Thanks!

moollaza commented on lib/DDG/Goodie/NoteFrequency.pm:
```diff
@@ -0,0 +1,88 @@
+    # must be a note letter, optional sharp or flat, octave number, and optional tuning frequency for A4
+    # e.g. "g#3 432"
+
+    if ( $_ =~ /^([A-Ga-g])([b#])?([0-8])(\s+[0-9]{1,5})?$/ ) {
```
nit: would remove the nesting and use ` return unless m/.../;`

moollaza commented on lib/DDG/Goodie/NoteFrequency.pm:
```diff
@@ -0,0 +1,88 @@
+        $pitchClass %= 12;
+
+        # validate note is between C0 and B8
+        if ( $midi >= 12 && $midi <= 119 ) {
```
same here, we usually use `return unless ...` which limits nesting

moollaza commented on t/NoteFrequency.t:
```diff
@@ -0,0 +1,40 @@
+        structured_answer => {
+            input     => ["C3 in A432 tuning"],
+            operation => "Note Frequency",
+            result    => "128.434368420294 Hz"
```
I'm not sure if it matters, but should we be rounding these numbers?

sublinear commented:
Thank you. Those are all good points. I have made the changes.

moollaza commented:
>Thank you. Those are all good points. I have made the changes.

@sublinear excellent, thanks! I've deployed the PR onto Beta.duckduckgo.com. Once I've given it a final review it should be good to go!

Thanks for doing an awesome job :+1:

mintsoft commented on lib/DDG/Goodie/NoteFrequency.pm:
```diff
@@ -0,0 +1,80 @@
+    # must be a note letter, optional sharp or flat, octave number, and optional tuning frequency for A4
+    # e.g. "g#3 432"
+
+    return unless ($_ =~ /^([A-Ga-g])([b#])?([0-8])(\s+[0-9]{1,4})?$/ );
```
can you add hz in an optional capture so that `notefreq c3 430hz` works?

moollaza commented on lib/DDG/Goodie/NoteFrequency.pm:
```diff
@@ -0,0 +1,80 @@
+use DDG::Goodie;
+use strict;
+
+use Math::Round;
```
Would you mind switching this to `Math::BigInt`? We don't have `Math::Round` installed on our production/dev machines, which means I'll need to have the team install it before I can merge this.

We *do* have Math::BigInt already though, so if you switch to that I should be able to deploy this to Beta and get it live sooner

sublinear commented:
Added a case-insensitive "hz" with optional preceding space to the regex expression, and I also just implemented the rounding myself. The gotcha being that it doesn't work with negative numbers, but the equation is never negative, so it's fine.

moollaza commented:
@sublinear this LGTM :+1:

Thanks a lot for contributing (we really appreciate it :smiley:). We invite you to stick around and comment/help on other PR's & Issues or even submit an Instant Answer!

moollaza commented:
By the way, we think you did a great job with this! Are you interested in helping others who build similar Instant Answers in the future?

If yes, we'll reach out and add you to a private list of mentors as well as send over some DDG merch as an additional thank you. Just email <open@duckduckgo.com> with the subject 'I want to help.' Please include your shipping address, phone-number (required for shipping outside USA) and t-shirt size. We'll reach out from there!

Otherwise, thank you for this amazing contribution. We hope you'll keep exploring and playing around with DuckDuckHack.

moollaza commented:
@sublinear congrats! It's live: https://duckduckgo.com/?q=note+frequency+of+c5&ia=answer

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's, tackle open issues or even submit more Instant Answers!

sublinear commented:
Very cool. Thanks for the support!
